pkg/skaffold/kubernetes/portforward/forwarder_manager.go

```diff
@@ -38,20 +43,69 @@ type ForwarderManager struct {
 }
 
 // NewForwarderManager returns a new port manager which handles starting and stopping port forwarding
-func NewForwarderManager(out io.Writer, cli *kubectl.CLI, podSelector kubernetes.PodSelector, namespaces []string, label string, opts config.PortForwardOptions, userDefined []*latest.PortForwardResource) *ForwarderManager {
+func NewForwarderManager(out io.Writer, cli *kubectl.CLI, podSelector kubernetes.PodSelector, namespaces []string, label string, runMode config.RunMode, options config.PortForwardOptions, userDefined []*latest.PortForwardResource) *ForwarderManager {
+	if !options.Enabled() {
+		return nil
+	}
+
+	// TODO this doesn't feel like the right place
+	if err := options.Validate(); err != nil {
+		logrus.Error("port-forward: ", err)
+		return nil
+	}
+
 	entryManager := NewEntryManager(out, NewKubectlForwarder(out, cli))
 
 	var forwarders []Forwarder
-	forwarders = append(forwarders, NewResourceForwarder(entryManager, label, userDefined))
-	if opts.ForwardPods {
-		forwarders = append(forwarders, NewWatchingPodForwarder(entryManager, podSelector))
+	if options.ForwardUser(runMode) {
+		forwarders = append(forwarders, NewUserDefinedForwarder(entryManager, userDefined))
+	}
+	if options.ForwardServices(runMode) {
+		forwarders = append(forwarders, NewServicesForwarder(entryManager, label))
+	}
+	if options.ForwardPods(runMode) {
+		forwarders = append(forwarders, NewWatchingPodForwarder(entryManager, podSelector, allPorts))
+	} else if options.ForwardDebug(runMode) {
+		forwarders = append(forwarders, NewWatchingPodForwarder(entryManager, podSelector, debugPorts))
 	}
 
 	return &ForwarderManager{
 		forwarders: forwarders,
 	}
 }
 
+func allPorts(pod *v1.Pod, c v1.Container) []v1.ContainerPort {
+	return c.Ports
+}
+
+func debugPorts(pod *v1.Pod, c v1.Container) []v1.ContainerPort {
+	var ports []v1.ContainerPort
+
+	annot, found := pod.ObjectMeta.Annotations[debugging.DebugConfigAnnotation]
```

Comment:
Not related to this PR, but curious about why are we using pod annotations to store debug configurations?
Is it only for IDEs? DO you know if IDES do use the metadata and if yes is can we instead stick debug metadata in the new proposed metadata API.

Reply:
It's there for tooling to consume.  Funnily enough, the major consumer is Skaffold itself which monitors for debuggable pods and turns them into `DebuggingContainerEvent`:

https://github.com/GoogleContainerTools/skaffold/blob/5bf0323473ce7610a6cde65ce1f882f886e029ea/pkg/skaffold/kubernetes/debugging/container_manager.go#L92-L93

(what's the new metadata api?)

Reply:
Sorry missed responding to this.  
re: whats is the new metadata API. 

For event API V2, there are some changes suggested by VSC to add additional metadata for driving the proposed new view.  Skaffold does have a MetaData or GetState endpoint which we are thinking of revamping.